File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.cs
using System;

public class Solution 
{
    public int solution(int[] ingredient) 
    {
        int[] stack = new int[ingredient.Length];
        int top = -1;  // 스택의 맨 위 인덱스 (-1이면 비어있음)
        int count = 0;
        
        foreach (int item in ingredient)
        {
            stack[++top] = item;  // push
            
            // 맨 위 4개가 [1,3,2,1] (위에서부터 빵,고기,야채,빵) 인지 확인
            if (top >= 3 
                && stack[top] == 1 
                && stack[top-1] == 3 
                && stack[top-2] == 2 
                && stack[top-3] == 1)
            {
                top -= 4;  // 4개 pop (그냥 인덱스만 4 줄이면 됨)
                count++;
            }
        }
        
        return count;
    }
}

File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

int solution(vector<int> ingredient) 
{
    string str = "";
    string hbg = "1231";
    
    int count = 0;
    for(int i=0; i<ingredient.size(); i++)
    {
        if(str.length() < 8)
        {
            str += to_string(ingredient[i]);
        }
        else if(str.length() >= 8)
        {
            str += to_string(ingredient[i]);
            str.erase(0,1);
        }
        
        while(str.find(hbg) != string::npos)
        {
            size_t pos = str.find(hbg);
            str.erase(pos, hbg.length());
            count++;
        }
    }
    return count;
}